x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameTransformTask.java

```diff
@@ -235,7 +235,7 @@ public long getInProgressCheckpoint() {
         }
     }
 
-    public void setTaskStateStopped() {
+    public synchronized void setTaskStateStopped() {
```

Comment:
I made this synchronized as `start` `stop` and `markAsFailed` are all synchronized and we do not want task state flipping between states while we are attempting to transition it. 

`doSaveState` calls `setTaskStateStopped` and since `synchronized` utilize reentrant locks, `stop` calling `doSaveState` directly in certain scenarios should not be a problem.